From 2fb141bf9c23cf605db2158dd7d6412f9854059e Mon Sep 17 00:00:00 2001
From: Guo Mengqi <guomengqi3@huawei.com>
Date: Tue, 22 Mar 2022 10:49:07 +0800
Subject: [PATCH 0234/1076] [Huawei] mm: sharepool: fix hugepage_rsvd count
 increase error

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5RO2H

--------------------------------

When nr_hugepages is configured, sharepool allocates hugepages first
from hugetlb pool, then from buddy system if the pool had been used up.
Current page release function treat the buddy system hugepages as
hugetlb pages, which caused HugePages_Rsvd to increase improperly.

Add a check in page release function:
    if the page is temporary, do not call hugetlb_unreserve_pages.

Signed-off-by: Guo Mengqi <guomengqi3@huawei.com>
---
 fs/hugetlbfs/inode.c | 19 +++++++++++++------
 mm/share_pool.c      |  3 +--
 2 files changed, 14 insertions(+), 8 deletions(-)

diff --git a/fs/hugetlbfs/inode.c b/fs/hugetlbfs/inode.c
index 8a87d1b43387..379aa008514d 100644
--- a/fs/hugetlbfs/inode.c
+++ b/fs/hugetlbfs/inode.c
@@ -597,11 +597,17 @@ static void remove_inode_hugepages(struct inode *inode, loff_t lstart,
 			 */
 			VM_BUG_ON(HPageRestoreReserve(page));
 			remove_huge_page(page);
-			freed++;
-			if (!truncate_op) {
-				if (unlikely(hugetlb_unreserve_pages(inode,
-							index, index + 1, 1)))
-					hugetlb_fix_reserve_counts(inode);
+			/*
+			 * if the page is from buddy system, do not add to freed.
+			 * because freed is used for hugetlbfs reservation accounting.
+			 */
+			if (!HPageTemporary(page)) {
+				freed++;
+				if (!truncate_op) {
+					if (unlikely(hugetlb_unreserve_pages(inode,
+									index, index + 1, 1)))
+						hugetlb_fix_reserve_counts(inode);
+				}
 			}
 
 			unlock_page(page);
@@ -1054,7 +1060,8 @@ static int hugetlbfs_error_remove_page(struct address_space *mapping,
 	pgoff_t index = page->index;
 
 	remove_huge_page(page);
-	if (unlikely(hugetlb_unreserve_pages(inode, index, index + 1, 1)))
+	if (!HPageTemporary(page) &&
+			unlikely(hugetlb_unreserve_pages(inode, index, index + 1, 1)))
 		hugetlb_fix_reserve_counts(inode);
 
 	return 0;
diff --git a/mm/share_pool.c b/mm/share_pool.c
index 95c90ed8af30..5e382ffc3ed5 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -4225,8 +4225,7 @@ vm_fault_t sharepool_no_page(struct mm_struct *mm,
 
 		page = alloc_huge_page(vma, haddr, 0);
 		if (IS_ERR(page)) {
-			page = alloc_huge_page_nodemask(hstate_file(vma->vm_file),
-						    node_id, NULL, GFP_KERNEL);
+			page = hugetlb_alloc_hugepage(node_id, HUGETLB_ALLOC_BUDDY);
 			if (!page)
 				page = ERR_PTR(-ENOMEM);
 		}
-- 
2.17.1

